datagen/src/main/scala/com/nvidia/rapids/tests/scaletest/TableGenerator.scala

```diff
@@ -0,0 +1,231 @@
+    - key group 3 with 3 columns (The unique count should be about 99% of the total number of rows).
+    - 10 data columns of various types
+    - 10 decimal(10, 2) columns with a relatively small value range
+   */
+  private def genBData(): DataFrame = {
+    val schema = "b_foreign_a long," +
+      "b_key3_1 string," +
+      "b_key3_2 date," +
+      "b_key3_3 long," +
+      (1 to 10).map(i => s"b_data_$i ${randomColumnType()}").mkString(", ") + ", " +
+      (1 to 10).map(i => s"b_data_small_value_range_$i decimal(10, 2)").mkString(", ")
+    val overlapRatio = 0.99
+    val bData = dbgen.addTable("b_data", schema, bNumRows)
+    // This column "Should overlap with a_facts.primary_a about 99% of the time"
+    bData("b_foreign_a").setSeedRange(1, Math.round(1 / overlapRatio * aNumRows))
+    (1 to 3).map(i => bData(s"b_key3_$i").setSeedRange(1, Math.round(bNumRows * 0.99)))
+    bData("b_key3_2").setValueRange(-719162, 2932896)
+    (1 to 10).map(i => bData(s"b_data_small_value_range_$i").setSeedRange(1, 10000))
+    bData.toDF(spark)
+  }
+
+  /**
+  c_data: Each scale factor corresponds to 100,000 rows.
+    - c_foreign_a Should overlap with a_facts.primary_a about 50% of the time
+    - key group 2 columns up to the complexity (each key should show up about 10 times)
+    - c_data_row_num_1 1 data column that is essentially the row number for order by tests in window
+    - 5 data columns
+    - 5 numeric data columns
+   */
+  private def genCData(): DataFrame = {
+    val schema = "c_foreign_a long," +
+      (1 to complexity).map(i => s"c_key2_$i ${randomColumnType()}").mkString(",") + "," +
+      "c_data_row_num_1 long," +
+      (1 to 5).map(i => s"c_data_$i ${randomColumnType()}").mkString(",") + "," +
+      (1 to 5).map(i => s"c_data_numeric_$i ${randomNumericColumnType()}").mkString(",")
+    val overlapRatio = 0.50
+    val cData = dbgen.addTable("c_data", schema, cNumRows)
+    cData("c_foreign_a").setSeedRange(1, Math.round(1 / overlapRatio * aNumRows))
+    (1 to complexity).map(i => cData(s"c_key2_$i").setSeedRange(1, 10))
+    cData("c_data_row_num_1").setValueRange(1, cNumRows)
+    cData.toDF(spark)
+  }
+  /**
+  d_data: Each scale factor corresponds to 100,000 rows.
+    - key group 2 complexity columns (each key should show up about 10 time, but the overlap with c_data for key group 2 should only be about 50%)
+    - 10 data columns
+   */
+  private def genDData(): DataFrame = {
+    val schema = (1 to complexity).map(i => s"d_key2_$i ${randomColumnType()}").mkString(",") +
+      "," +
+      (1 to 10).map(i => s"d_data_$i ${randomColumnType()}").mkString(",")
+    val dData = dbgen.addTable("d_data", schema, dNumRows)
+    // each key should show up about 10 time, but the overlap with c_data for key group 2 should only be about 50%
+    // c_data key group uses seed range (1, 10) so here d_data uses (6, 15) to get 50% overlap
+    (1 to complexity).map(i => dData(s"d_key2_$i").setSeedRange(6, 15))
```

Comment:
We need to use the keygroup API specifically with correlated key group to actually make keygroup 2 behave correctly. The seed range calculation is fine, but because the keys are not correlated to each other we will not get the desired behavior.